### DISCUSSION TITLE
Interpolation error(?) with zero translation in Affine

### DISCUSSION THREAD

**Discussion by ntatsisk** [Project-MONAI/MONAI#5614] [Q&A]
Hello,

Doing an `Affine` transform with `translate_params=0` leads some kind of interpolation error between the input and the output tensor, which in theory should be identical. When setting `mode="nearest"` there is no such error as expected. 

Code to reproduce:
```
import torch
from monai.transforms import Affine

def calculate_max_error(tensor_shape, iterations=1000):
    # Zero translation
    affine_transform = Affine(translate_params=0, padding_mode="zeros")

    # Calculate max absolute error between input_tensor and output_tensor
    errors = []
    for i in range(iterations):
        input_tensor = 1000*torch.rand(tensor_shape)
        output_tensor, output_affine = affine_transform(input_tensor)
        diff = output_tensor - input_tensor
        errors.append(diff.abs().max())

    return torch.tensor(errors).max().item()

def print_error_for_shape(tensor_shape):
    print("Shape: {} - error: {}".format(tensor_shape, calculate_max_error(tensor_shape)))

print_error_for_shape((1, 2, 2, 2))
print_error_for_shape((1, 1, 1, 3))
print_error_for_shape((1, 3, 3, 3))
print_error_for_shape((1, 1, 1, 4))
print_error_for_shape((1, 4, 4, 4))
print_error_for_shape((1, 8, 8, 8))
print_error_for_shape((1, 9, 9, 9))
print_error_for_shape((1, 25, 25, 25))
print_error_for_shape((1, 27, 27, 27))
print_error_for_shape((1, 50, 50, 50))
```

which gives the following output:
```
Shape: (1, 2, 2, 2) - error: 0.0
Shape: (1, 1, 1, 3) - error: 0.0
Shape: (1, 3, 3, 3) - error: 0.0
Shape: (1, 1, 1, 4) - error: 6.103515625e-05
Shape: (1, 4, 4, 4) - error: 0.00018310546875
Shape: (1, 8, 8, 8) - error: 0.00030517578125
Shape: (1, 9, 9, 9) - error: 0.0
Shape: (1, 25, 25, 25) - error: 0.0013427734375
Shape: (1, 27, 27, 27) - error: 0.005126953125
Shape: (1, 50, 50, 50) - error: 0.01068115234375
```

Is this normal? Why is this interpolation artifact created in the first place? Equally puzzling: why some of the cases give zero error?

I am using:
```
MONAI version: 1.1.dev2248
Numpy version: 1.23.5
Pytorch version: 1.13.0+cu117
MONAI flags: HAS_EXT = False, USE_COMPILED = False, USE_META_DICT = False
MONAI rev id: 3400bd91422ccba9ccc3aa2ffe7fecd4eb5596bf
```

Thank you!

--
xref for our convenience: https://github.com/InsightSoftwareConsortium/ITKElastix/issues/126

**Comment by wyli** [comment]
thanks for reporting, with float64 it'll be more precise

```py
affine_transform = Affine(translate_params=0, padding_mode="zeros", dtype=torch.float64)
```

```
Shape: (1, 2, 2, 2) - error: 0.0
Shape: (1, 1, 1, 3) - error: 0.0
Shape: (1, 3, 3, 3) - error: 0.0
Shape: (1, 1, 1, 4) - error: 0.0
Shape: (1, 4, 4, 4) - error: 0.0
Shape: (1, 8, 8, 8) - error: 0.0
Shape: (1, 9, 9, 9) - error: 0.0
Shape: (1, 25, 25, 25) - error: 5.949706330908577e-13
Shape: (1, 27, 27, 27) - error: 0.0
Shape: (1, 50, 50, 50) - error: 6.532264746478278e-12
```

**Comment by ntatsisk** [comment]
Thanks @wyli for the reply! The solution that you provided indeed addresses the problem for zero translation. However, when I eventually start using non-zero values I notice higher interpolation error compared to e.g. ITK resampling. I might be wrong though. Is it fair to say that when `normalized=False` (default value) then `Affine` does actually two consecutive interpolations? One for the "coordinate normalization" and one for requested transformation/resampling?

Any thought or resource/link will help me a lot because I am working on an affine matrix bridge between ITK and MONAI.

**Comment by wyli** [comment]
that's an interesting question, coordinate normalization is remapping it to -1 and 1 so that the pytorch resampling backend can understand it

I made some tests comparing monai and itkpython
https://github.com/Project-MONAI/MONAI/blob/e2707418c1d41ff06c7cddb0b98ca977720d4c87/tests/test_meta_affine.py#L163-L175

another related unit tests is https://github.com/Project-MONAI/MONAI/blob/99b9dd7cf0667113e56f0dd90d40699115048d59/tests/test_warp.py#L110

please let me know if you have specific code examples for your questions, I'm happy to look into them.
